lib/providers/auth/app_initialization_provider.dart

```diff
@@ -13,31 +13,59 @@ final retryCounterProvider = StateProvider<int>((ref) => 0);
 
 final appInitializationProvider =
     FutureProvider.family<void, BuildContext>((ref, context) async {
+  var _authProvider = ref.read(authProvider);
   try {
-    var _authProvider = ref.read(authProvider);
+    var retryCount = ref.read(retryCounterProvider);
+    if (retryCount < 2) {
+      _checkInternetConnectivity(context, ref);
+    }
     await _authProvider.initializeUser();
-    var _user = _authProvider.userRes.body as UserTokenModel;
-    _assignNewTokenToDioAndAudioPlayer(ref, _user.token);
-    var deviceInfo = await ref.read(deviceAndAppInfoProvider.future);
-    var data = _setAppOpenedModelData(deviceInfo);
+    var res = _authProvider.userRes;
+    if (res.body != null) {
+      var _user = res.body as UserTokenModel;
+      _assignNewTokenToDioAndAudioPlayer(ref, _user.token);
+      var deviceInfo = await ref.read(deviceAndAppInfoProvider.future);
+      var data = _setAppOpenedModelData(deviceInfo);
 
-    await ref.read(eventsProvider(event: data.toJson()).future);
-    context.go(RouteConstants.homePath);
+      await ref.read(eventsProvider(event: data.toJson()).future);
+      ref.read(retryCounterProvider.notifier).update((state) => 0);
+      context.go(RouteConstants.homePath);
+    } else {
+      _handleRetry(context, ref, false);
+    }
   } catch (err) {
-    print(err);
-    _handleRetry(context, ref, err.toString());
+    _handleRetry(context, ref, _authProvider.userRes.body != null);
   }
 });
 
-void _handleRetry(BuildContext context, Ref ref, String err) {
+void _handleRetry(BuildContext context, Ref ref, bool shouldNavigate) {
   var retryCount = ref.read(retryCounterProvider);
-  if (retryCount < 4) {
-    Future.delayed(Duration(seconds: 2), () {
-      ref.refresh(appInitializationProvider(context));
-    });
+  if (retryCount < 2) {
+    _recallProvider(ref, context);
     ref.read(retryCounterProvider.notifier).update((state) => ++state);
   } else {
-    context.go(RouteConstants.downloadsPath);
+    if (shouldNavigate) {
+      ref.read(retryCounterProvider.notifier).update((state) => 0);
+      context.go(RouteConstants.downloadsPath);
+    } else {
+      _recallProvider(ref, context);
+      if (retryCount < 2) {
+        showSnackBar(context, StringConstants.timeout);
+      }
```

Comment:
This will never be reached, will it? You're saying 

```
if( retry < 2 ){
} else {
 if(retry < 2){ 
 }
}
```

Reply:
I also noticed in auth_provider generateUserToken, notifyListeners is being called twice. @osamaasifoak